### ISSUE TITLE
Large Splashes Can't Generate Schedules PDF

### ISSUE THREAD

**Issue by jmoldow** [learning-unlimited/ESP-Website#1198]
In spring 2014, at least two large Splash programs (Clark and Stanford) had trouble generating the student schedules.
- The Clark .pdf file was not downloadable from the website. I was able to download the .tex file from the website, and use that to manually create the .pdf file on the server using `gen_latex()`.
- Neither the Stanford .pdf file nor .tex file were downloadable from the website. I was able to manually create the .tex file on the server, and was also able to generate the .tex file by running `render_to_latex`. The .pdf file was also able to be generated this way, but it took a very long time. The bottleneck was running `ps2pdf`.

Stanford has been running large programs for a while now, and I don't think Clark's Splash is the biggest non-Stanford Splash ever, so I don't know why this is happening now. Maybe we changed something that made these files larger even if the content isn't larger. Or maybe we reconfigured diogenes wrong when we reinstalled the OS, and it can't handle large files as well.

**Comment by benjaminjkraft** [comment]
Putting here what I put in email: I think at least some of the issues with things getting stuck between generating TeX and returning a PDF to the user come from `dvips`, and thus would be fixed by the change to `pdflatex` in #1509.  Probably not the only issue, but it's at least one that should just require a backport to solve.

**Comment by willgearty** [comment]
Update from Stanford:
I haven't been able to generate a pdf of student schedules via the website for the past 2 years (it usually takes 10+ minutes, then crashes).
I've resorted to generating the latex code and then compiling that locally.

**Comment by benjaminjkraft** [comment]
Hmm, maybe we have a new issue here then!  It'll take some more investigation to see why things are breaking now; I think all the information above was a different issue fixed in #1759, as it turned out.

**Comment by mariolsen0112358** [comment]
MIT Splash 2016 had issues downloading the schedules PDF (on Windows and Mac) with the same 10+ minutes load -> crash issue. However, Favyen managed it first try without issues on whatever (Linux?) he uses and said he saw "some connection refused errors from varnish in the log" after our failed attempts.

There was some speculation that it might be due to a special character on one of the schedules? Or just sheer size of the doc. In any case, smaller groups ("Students with Saturday dinner" as opposed to "Students who filled out medliab") loaded fine.

**Comment by willgearty** [comment]
Yeah, we haven't had trouble downloading smaller subsets of the student schedules, either.

Oddly, I just tried downloading the schedules (_without barcodes_, ~2000 pages) with 3 different browsers on Windows (Chrome, Waterfox, and Edge) and both Edge and Waterfox downloaded them just fine, but Chrome got a 500 Internal Error.
_With barcodes_ (~4000 pages), Waterfox got a 500 Internal Error, Chrome finished loading the request but gives a "Failed to load PDF document" error, and Edge successfully downloaded the file to disk.

Perhaps it's just a luck of the draw thing? Or perhaps this could be partly related to #1769?

**Comment by Supriyo760** [comment]
Hi @willgearty and team,

Looking closely at the symptoms described here over the years (500 Internal Server Errors after 10+ minutes, "connection refused" from Varnish, and intermittent browser-specific crashes), this exhibits the classic signs of an upstream HTTP timeout. Reverse proxies (like Varnish or Nginx) and web browsers are specifically configured to drop connections if the backend server takes more than a few minutes to respond.

Because compiling thousands of pages of LaTeX into a PDF is a highly CPU-bound and time-intensive operation, forcing the main Django web worker thread to block for 15+ minutes is an architectural bottleneck. It not only starves other incoming requests by tying up a worker but guarantees network timeouts for large programs.

Proposed Solution: We need to decouple the PDF compilation from the synchronous HTTP request/response cycle. This ties into the concept mentioned in #3454.

Background Processing: When an admin requests a large printable (like all student schedules), the view should instantly dispatch the generation job to an asynchronous background task worker (e.g., Celery, or whatever async queue is configured for ESP-Website).
Persistent Storage: The background worker computes the .tex, compiles the .pdf, and saves the resulting file to a designated media directory or a database FileField.
UX Handoff: The admin is immediately redirected to a "Generation in Progress" status page that periodically polls for completion. Alternatively, we simply email the admin a secure download link once the background task finishes.
This completely eliminates the browser timeout crashes, prevents web workers from hanging, and stabilizes server load.

If this hasn't been implemented yet and you'd like to modernize the printables architecture, I would love to take this on! Could I be assigned to this issue?

**Comment by willgearty** [comment]
That sounds like a great solution, thanks @Supriyo760. I think a page that shows the status would be good.
